### PR TITLE
PP-885/explore-solution-b2

### PR DESCRIPTION
**Pull request by franciscotobar**
## What

- Implements solution [B2](https://docs.google.com/document/d/10N8jV41KwOXdoiCcSGHRvS5TYGyA_1KkVpee75eG3Gg/edit#heading=h.ddb8zsm0g2vy)

## Why

- Necessary to execute a contract while deploying a smart wallet

## Refs

-  [PP-885](https://rsklabs.atlassian.net/browse/PP-885)

### CODE REVIEW

File: test/factory/smartWalletFactory.test.ts
Comment (antomor) [inline test/factory/smartWalletFactory.test.ts:1]: if possible, I'd add some tests to check that the error is forwarded

File: test/smartwallet/smartWallet.test.ts [hunk 1]
```diff
@@ -175,18 +179,242 @@ describe('SmartWallet contract', function () {
...
+
+        const ABI = ['function isInitialized()'];
+        const abiInterface = new hardhat.utils.Interface(ABI);
+        recipientFunction = abiInterface.encodeFunctionData(
+          'isInitialized',
+          []
+        );
+      });
+
+      it('Should pay on not sponsored(token) deployment', async function () {
```
Comment (antomor) [inline test/smartwallet/smartWallet.test.ts:213]:
```suggestion
      it('Should pay for deployment using token', async function () {
```

File: test/smartwallet/smartWallet.test.ts [hunk 2]
```diff
@@ -175,18 +179,242 @@ describe('SmartWallet contract', function () {
...
+            worker.address,
+            signature
+          );
+
+        expect(fakeToken.transfer).to.be.called;
+        expect(recipient.isInitialized, 'Recipient method was not called').to.be
+          .called;
+      });
+
+      it('Should pay on not sponsored(native) deployment', async function () {
```
Comment (antomor) [inline test/smartwallet/smartWallet.test.ts:259]:
```suggestion
      it('Should pay for deployment using native', async function () {
```

File: test/smartwallet/smartWallet.test.ts [hunk 3]
```diff
@@ -175,18 +179,242 @@ describe('SmartWallet contract', function () {
...
+            .connect(relayHub)
+            .relayedUserSmartWalletCreation(
+              deployRequest.request,
+              suffixData,
+              worker.address,
+              signature
+            )
+        ).to.be.rejectedWith('Unable to execute');
+
+        expect(recipient.isInitialized, 'Recipient method was not called').to.be
```
Comment (antomor) [inline test/smartwallet/smartWallet.test.ts:412]: I'd also check that the revert message corresponds to `Failing in purpose`

File: test/smartwallet/smartWallet.test.ts [hunk 4]
```diff
@@ -217,14 +445,66 @@ describe('SmartWallet contract', function () {
...
+          deployRequest.request,
+          suffixData,
+          worker.address,
+          signature
+        );
+
+      const balanceAfter = await provider.getBalance(smartWalletAddress);
+
+      expect(balanceBefore).to.be.equal(amountToBePaid);
+      expect(balanceAfter).to.be.equal(balanceBefore.sub(amountToBePaid));
```
Comment (antomor) [inline test/smartwallet/smartWallet.test.ts:504]: I'd check for the receiver balance rather than the SW one

**Review by antomor — CHANGES_REQUESTED** [review]
@franciscotobar I've left some minor comments in the PR

File: contracts/smartwallet/SmartWallet.sol [hunk 5]
```diff
@@ -270,22 +270,43 @@ contract SmartWallet is IForwarder {
...
+        uint256 value,
+        uint256 executionGas,
+        bytes calldata data
+    ) external returns (bool success, bytes memory ret) {
+        require(getOwner() == bytes32(0), "Already initialized");
 
         _setOwner(owner);
+        if (to != address(0)) {
+            require(gasleft() > executionGas, "Not enough gas left");
+            (success, ret) = to.call{gas: executionGas, value: value}(data);
```
[the remark this answers is not in the document]
Comment (franciscotobar) [reply contracts/smartwallet/SmartWallet.sol:284]: are we sure that we want to do this validation `on-chain`, the server will not execute anything that is not being validated by the verifiers.

File: contracts/factory/SmartWalletFactory.sol [hunk 6]
```diff
@@ -233,10 +241,14 @@ contract SmartWalletFactory is ISmartWalletFactory {
         //required is done via the runtime code, to avoid the parameters impacting on the resulting address
 
         /* solhint-disable-next-line avoid-low-level-calls */
-        (bool success, ) = addr.call(initdata);
+        (bool success, bytes memory ret) = addr.call(initdata);
 
         /* solhint-disable-next-line reason-string */
```
[the remark this answers is not in the document]
Comment (franciscotobar) [reply contracts/factory/SmartWalletFactory.sol:246]: updated

File: contracts/factory/SmartWalletFactory.sol [hunk 7]
```diff
@@ -134,12 +134,16 @@ contract SmartWalletFactory is ISmartWalletFactory {
 
         //a6b63eb8  =>  initialize(address owner,address tokenAddr,address tokenRecipient,uint256 tokenAmount,uint256 tokenGas)
```
[the remark this answers is not in the document]
Comment (franciscotobar) [reply contracts/factory/SmartWalletFactory.sol:135]: updated

File: contracts/smartwallet/SmartWallet.sol [hunk 8]
```diff
@@ -270,22 +270,43 @@ contract SmartWallet is IForwarder {
         address tokenAddr,
         address tokenRecipient,
         uint256 tokenAmount,
-        uint256 tokenGas
-    ) external {
-        require(getOwner() == bytes32(0), "already initialized");
+        uint256 tokenGas,
+        address to,
+        uint256 value,
+        uint256 executionGas,
```
[the remark this answers is not in the document]
Comment (franciscotobar) [reply contracts/smartwallet/SmartWallet.sol:276]: updated

File: test/factory/smartWalletFactory.test.ts
Comment (franciscotobar) [reply test/factory/smartWalletFactory.test.ts:1]: updated

File: test/smartwallet/smartWallet.test.ts (hunk 1 above)
Comment (franciscotobar) [reply test/smartwallet/smartWallet.test.ts:213]: updated

File: test/smartwallet/smartWallet.test.ts (hunk 2 above)
Comment (franciscotobar) [reply test/smartwallet/smartWallet.test.ts:259]: updated

File: test/smartwallet/smartWallet.test.ts (hunk 3 above)
Comment (franciscotobar) [reply test/smartwallet/smartWallet.test.ts:412]: updated

File: test/smartwallet/smartWallet.test.ts (hunk 4 above)
Comment (franciscotobar) [reply test/smartwallet/smartWallet.test.ts:504]: updated

File: contracts/smartwallet/SmartWallet.sol [hunk 9]
```diff
@@ -174,10 +160,35 @@ contract SmartWallet is IForwarder {
...
         //methods that revert if the gasleft() is not enough to execute whatever logic they have.
-
         require(gasleft() > req.gas, "Not enough gas left");
         (success, ret) = req.to.call{gas: req.gas, value: req.value}(req.data);
 
+        if (req.tokenAmount > 0) {
+            bool successPayment;
+            bytes memory retPayment;
+            if (req.tokenContract == address(0)) {
+                (successPayment, retPayment) = payable(feesReceiver).call{
```
Comment (raullaprida) [inline contracts/smartwallet/SmartWallet.sol:170]: Why are you using a call to transfer RBTC instead of a transfer? payable(feesReceiver).transfer(balanceToTransfer); ?

File: contracts/smartwallet/SmartWallet.sol
Comment (raullaprida) [inline contracts/smartwallet/SmartWallet.sol:1]: I wouldn't modify the original template. I would leave the template as is, and I would add a new template with this extra functionality for whoever wants this new functionality (which is only Boltz so far, the rest of the RIF Relay audience does not care about these new scenarios)

**Review by raullaprida — CHANGES_REQUESTED** [review]
The review is not finished, but I wanted to share it with you because Im proposing creating a new template for this scenario instead of modifying the existing one, and that is a big change

File: contracts/RelayHub.sol [hunk 10]
```diff
@@ -164,11 +164,13 @@ contract RelayHub is IRelayHub {
...
-        bool deploySuccess;
         bytes memory ret;
-        (deploySuccess, ret) = Eip712Library.deploy(deployRequest, signature);
+        (destinationCallSuccess, ret) = Eip712Library.deploy(
+            deployRequest,
+            signature
+        );
 
-        if (!deploySuccess) {
+        if (!destinationCallSuccess) {
```
[the remark this answers is not in the document]
Comment (franciscotobar) [reply contracts/RelayHub.sol:173]: I left it here because I wanted to discuss with you, Im going to ask you the question in the chat about the reasoning.

File: contracts/smartwallet/SmartWallet.sol [hunk 11]
```diff
@@ -122,9 +122,10 @@ contract SmartWallet is IForwarder {
         (success, ret) = to.call{value: msg.value}(data);
 
         //If any balance has been added then trasfer it to the owner EOA
-        if (address(this).balance > 0) {
+        uint256 balanceToTransfer = address(this).balance;
```
[the remark this answers is not in the document]
Comment (franciscotobar) [reply contracts/smartwallet/SmartWallet.sol:125]: I did not do this comparison, but Im going to check it and standardize it across the smart wallet.

File: contracts/smartwallet/SmartWallet.sol (hunk 9 above)
Comment (franciscotobar) [reply contracts/smartwallet/SmartWallet.sol:170]: According to [solidity by example ](https://solidity-by-example.org/sending-ether/)this is the recommended way.

File: contracts/smartwallet/SmartWallet.sol
Comment (franciscotobar) [reply contracts/smartwallet/SmartWallet.sol:1]: Ill be moving the DeployVerifier aswell and creating a new RelayVerifier for boltz.

File: contracts/smartwallet/SmartWallet.sol (hunk 11 above)
[the remark this answers is not in the document]
Comment (franciscotobar) [reply contracts/smartwallet/SmartWallet.sol:125]: I just check it, and you are right, it consumes 4 more of gas, Ill be standardizing it.

File: contracts/RelayHub.sol (hunk 10 above)
[the remark this answers is not in the document]
Comment (franciscotobar) [reply contracts/RelayHub.sol:172]: updated

File: contracts/smartwallet/SmartWallet.sol (hunk 11 above)
[the remark this answers is not in the document]
Comment (franciscotobar) [reply contracts/smartwallet/SmartWallet.sol]: updated

File: contracts/smartwallet/SmartWallet.sol [hunk 12]
```diff
@@ -174,10 +160,35 @@ contract SmartWallet is IForwarder {
         //not enough, in that case there might be enough gas to complete the relay and the token payment would be collected
         //For that reason, the next require line must be left uncommented, to avoid malicious relayer attacks to destination contract
         //methods that revert if the gasleft() is not enough to execute whatever logic they have.
-
         require(gasleft() > req.gas, "Not enough gas left");
```
[the remark this answers is not in the document]
Comment (franciscotobar) [reply contracts/smartwallet/SmartWallet.sol]: updated

File: contracts/smartwallet/SmartWallet.sol
Comment (franciscotobar) [reply contracts/smartwallet/SmartWallet.sol:1]: updated

File: contracts/smartwallet/BoltzSmartWallet.sol [hunk 13]
```diff
@@ -0,0 +1,330 @@
...
+     * @param to - Destination contract to execute
+     * @param value - Value to send to destination contract
+     * @param destinationCallGas - Gas limit of destination contract
+     * @param data - Data to be execute by destination contract
+     */
+    function initialize(
+        address owner,
+        address tokenContract,
+        address feesReceiver,
+        uint256 tokenAmount,
```
Comment (raullaprida) [inline contracts/smartwallet/BoltzSmartWallet.sol:275]: The initialize function is not part of an Interface. The name and number of attributes could be reshaped to the Boltz Scenario. In Boltz, a user does not pay with Tokens. If we want to have an option that has all the possible scenarios (pay with tokens, or RBTC), that's fine but then we need another template that is minimalistic and only implements what Boltz is requiring (the deploy of the smart wallet is paid with the RBTC from the claim function). Then they can choose whatever template fits them best (they could, for example, choose the minimalistic one if it is much cheaper to execute, as it requires less parameters and less opcodes)

File: contracts/verifier/BoltzDeployVerifier.sol [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+                relayRequest.request.recoverer,
+                relayRequest.request.index
+            );
+
+        require(!_isContract(contractAddr), "Address already created!");
+
+        if (relayRequest.request.tokenAmount > 0) {
+            if (relayRequest.request.tokenContract != address(0)) {
+                require(
+                    tokens[relayRequest.request.tokenContract],
```
Comment (raullaprida) [inline contracts/verifier/BoltzDeployVerifier.sol:57]: Same comment as before, in the minimalistic scenario, Boltz would not use ERC20 tokens

File: contracts/verifier/BoltzRelayVerifier.sol
Comment (raullaprida) [inline contracts/verifier/BoltzRelayVerifier.sol:1]: In the minimalistic version for Boltz (transient), they wouldn't use this verifier. Just to take into consideration, that this is only for the full-featured template

File: contracts/factory/CustomSmartWalletFactory.sol [hunk 15]
```diff
@@ -298,7 +298,7 @@ contract CustomSmartWalletFactory is ICustomSmartWalletFactory {
         return
             abi.encodePacked(
                 keccak256(
-                    "RelayRequest(address relayHub,address from,address to,address tokenContract,address recoverer,uint256 value,uint256 nonce,uint256 tokenAmount,uint256 tokenGas,uint256 validUntilTime,uint256 index,bytes data,RelayData relayData)RelayData(uint256 gasPrice,address feesReceiver,address callForwarder,address callVerifier)"
+                    "RelayRequest(address relayHub,address from,address to,address tokenContract,address recoverer,uint256 value,uint256 gas,uint256 nonce,uint256 tokenAmount,uint256 tokenGas,uint256 validUntilTime,uint256 index,bytes data,RelayData relayData)RelayData(uint256 gasPrice,address feesReceiver,address callForwarder,address callVerifier)"
```
Comment (raullaprida) [inline contracts/factory/CustomSmartWalletFactory.sol:301]: Is this related to this new functionality? I ask because this is the generic Custom Factory, not the Boltz one

File: contracts/factory/SmartWalletFactory.sol [hunk 16]
```diff
@@ -261,7 +261,7 @@ contract SmartWalletFactory is ISmartWalletFactory {
         return
             abi.encodePacked(
                 keccak256(
-                    "RelayRequest(address relayHub,address from,address to,address tokenContract,address recoverer,uint256 value,uint256 nonce,uint256 tokenAmount,uint256 tokenGas,uint256 validUntilTime,uint256 index,bytes data,RelayData relayData)RelayData(uint256 gasPrice,address feesReceiver,address callForwarder,address callVerifier)"
+                    "RelayRequest(address relayHub,address from,address to,address tokenContract,address recoverer,uint256 value,uint256 gas,uint256 nonce,uint256 tokenAmount,uint256 tokenGas,uint256 validUntilTime,uint256 index,bytes data,RelayData relayData)RelayData(uint256 gasPrice,address feesReceiver,address callForwarder,address callVerifier)"
```
Comment (raullaprida) [inline contracts/factory/SmartWalletFactory.sol:264]: Can you please explain how this change in the General Factory is required for Boltz?

File: contracts/smartwallet/MinimalBoltzSmartWallet.sol
Comment (raullaprida) [inline contracts/smartwallet/MinimalBoltzSmartWallet.sol:1]: This is a Smart Wallet template that doesn't allow ERC20 payments, it's not really the minimalistic version for Boltz. The minimalistic version for Boltz is a wallet that is used once to claim the RBTC that is sent to the owner EOA of that Smart Wallet. That means, this wallet doesn't really need to have execute functionalities, nor recovery, or anything like that. All of this can be left on the BoltzSmartWallet template. This new minimalistic template only initialises the SmartWallet and sends the RBTC to its owner,  the rest is blank, since IForwarder is forcing you to implement some functions you have two alternatives:
1: You don't inherit from IForwarder for this SWallet template. You can create a new interface for the functions you do need to successfully initialise the SmartWallet (getOwner basically)
2: You can still inherit from IForwarder and leave all those functions empty (a bit ugly)

The idea of this SmartWallet is to be the transtient option, only usable once and then discarded.  We have the BoltzSmartWallet as the option with all the SmartWallet features PLUS enabling the option to call the claim contract

**Review by raullaprida — CHANGES_REQUESTED** [review]
I think the minimalistic approach is not the single use smart wallet version we discussed

**Review by raullaprida — CHANGES_REQUESTED** [review]
My comments are minimal, this PR is almost done!

File: contracts/smartwallet/BoltzSmartWallet.sol [hunk 17]
```diff
@@ -0,0 +1,330 @@
...
+        address to,
+        uint256 value,
+        uint256 destinationCallGas,
+        bytes calldata data
+    ) external returns (bool success, bytes memory ret) {
+        require(getOwner() == bytes32(0), "Already initialized");
+
+        _setOwner(owner);
+        if (to != address(0)) {
+            (success, ret) = to.call{gas: destinationCallGas, value: value}(
```
[the remark this answers is not in the document]
Comment (franciscotobar) [reply contracts/smartwallet/BoltzSmartWallet.sol]: updated

File: contracts/smartwallet/BoltzSmartWallet.sol [hunk 18]
```diff
@@ -0,0 +1,330 @@
...
+
+        //we need to initialize the contract
+        if (tokenAmount > 0) {
+            if (tokenContract == address(0)) {
+                (success, ret) = payable(feesReceiver).call{
+                    value: tokenAmount,
+                    gas: tokenGas
+                }("");
+            } else {
+                (success, ret) = tokenContract.call{gas: tokenGas}(
```
[the remark this answers is not in the document]
Comment (franciscotobar) [reply contracts/smartwallet/BoltzSmartWallet.sol:301]: updated

File: contracts/smartwallet/BoltzSmartWallet.sol [hunk 19]
```diff
@@ -0,0 +1,330 @@
...
+                if (ret.length == 0) revert("Unable to execute");
+                assembly {
+                    revert(add(ret, 32), mload(ret))
+                }
+            }
+        }
+
+        //we need to initialize the contract
+        if (tokenAmount > 0) {
+            if (tokenContract == address(0)) {
```
[the remark this answers is not in the document]
Comment (franciscotobar) [reply contracts/smartwallet/BoltzSmartWallet.sol:295]: updated

File: contracts/smartwallet/BoltzSmartWallet.sol (hunk 13 above)
Comment (franciscotobar) [reply contracts/smartwallet/BoltzSmartWallet.sol:274]: updated

File: contracts/smartwallet/BoltzSmartWallet.sol
[the remark this answers is not in the document]
Comment (franciscotobar) [reply contracts/smartwallet/BoltzSmartWallet.sol:1]: updated

File: contracts/verifier/BoltzRelayVerifier.sol
Comment (franciscotobar) [reply contracts/verifier/BoltzRelayVerifier.sol:1]: updated

File: contracts/factory/CustomSmartWalletFactory.sol (hunk 15 above)
Comment (franciscotobar) [reply contracts/factory/CustomSmartWalletFactory.sol]: updated

File: contracts/factory/SmartWalletFactory.sol (hunk 16 above)
Comment (franciscotobar) [reply contracts/factory/SmartWalletFactory.sol]: updated

File: contracts/smartwallet/MinimalBoltzSmartWallet.sol
Comment (franciscotobar) [reply contracts/smartwallet/MinimalBoltzSmartWallet.sol:1]: updated

File: test/factory/minimalBoltzSmartWalletFactory.test.ts [hunk 20]
```diff
@@ -0,0 +1,572 @@
...
+              signature
+            )
+        ).to.be.rejectedWith('Unable to pay for deployment');
+
+        const finalReceiverBalance = await feesReceiver.getBalance();
+
+        expect(finalReceiverBalance).to.be.equal(initialReceiverBalance);
+      });
+
+      it('should fail when the smart wallet does not have funds to pay using token', async function () {
```
[the remark this answers is not in the document]
Comment (franciscotobar) [reply test/factory/minimalBoltzSmartWalletFactory.test.ts:215]: updated

File: test/smartwallet/boltzSmartWallet.template.test.ts
[the remark this answers is not in the document]
Comment (franciscotobar) [reply test/smartwallet/boltzSmartWallet.template.test.ts:1]: No, a new one was added

File: contracts/verifier/MinimalBoltzRelayVerifier.sol [hunk 21]
```diff
@@ -0,0 +1,57 @@
...
+
+    /* solhint-disable no-unused-vars */
+    function verifyRelayedCall(
+        EnvelopingTypes.RelayRequest calldata relayRequest,
+        bytes calldata signature
+    ) external virtual override returns (bytes memory context) {
+        address payer = relayRequest.relayData.callForwarder;
+
+        // Check for the codehash of the smart wallet sent
+        bytes32 smartWalletCodeHash;
```
[the remark this answers is not in the document]
Comment (franciscotobar) [reply contracts/verifier/MinimalBoltzRelayVerifier.sol:45]: @antomor can you please provide your feedback regarding this :)

File: contracts/smartwallet/MinimalBoltzSmartWallet.sol [hunk 22]
```diff
@@ -0,0 +1,65 @@
...
+     * @param feesReceiver - Recipient of payment
+     * @param tokenAmount - Amount to pay
+     * @param tokenGas - Gas limit of payment
+     * @param to - Destination contract to execute
+     * @param data - Data to be execute by destination contract
+     */
+    function initialize(
+        address owner,
+        address feesReceiver,
+        uint256 tokenAmount,
```
[the remark this answers is not in the document]
Comment (franciscotobar) [reply contracts/smartwallet/MinimalBoltzSmartWallet.sol]: updated

File: contracts/verifier/MinimalBoltzRelayVerifier.sol (hunk 21 above)
[the remark this answers is not in the document]
Comment (antomor) [reply contracts/verifier/MinimalBoltzRelayVerifier.sol:45]: @franciscotobar I agree, I think we could just revert. No need to have additional checks there

File: contracts/verifier/MinimalBoltzDeployVerifier.sol [hunk 23]
```diff
@@ -0,0 +1,96 @@
...
+            );
+
+            BoltzTypes.ClaimInfo memory claim = abi.decode(
+                relayRequest.request.data[4:],
+                (BoltzTypes.ClaimInfo)
+            );
+
+            require(
+                relayRequest.request.tokenAmount <= claim.amount,
+                "Native balance too low"
```
Comment (antomor) [inline contracts/verifier/MinimalBoltzDeployVerifier.sol:71]: I think this message may be misleading so I'd change to something like `Claiming value lower than fees` or `Claiming value does not cover for the server fees`

Comment (franciscotobar) [reply contracts/verifier/MinimalBoltzDeployVerifier.sol:71]: updated

File: hardhat.config.ts [hunk 24]
```diff
@@ -166,6 +178,35 @@ task('remove-tokens', 'Removes a list of tokens')
     await removeTokens(taskArgs, hre);
   });
 
+task('allow-contracts', 'Allows a list of contracts')
+  .addParam('contractList', 'list of contracts')
```
Comment (antomor) [inline hardhat.config.ts:182]:
```suggestion
  .addParam('contractList', 'list of contracts in a comma-separated format (e.g.: "address1,address2")')
```

File: hardhat.config.ts [hunk 25]
```diff
@@ -166,6 +178,35 @@ task('remove-tokens', 'Removes a list of tokens')
...
+  .addOptionalParam(
+    'verifierList',
+    'list of contracts in a comma-separated format (e.g.: "address1,address2")'
```
Comment (antomor) [inline hardhat.config.ts:185]:
```suggestion
    'list of verifiers in a comma-separated format (e.g.: "address1,address2")'
```

File: tasks/destinationContractHandler/allowContracts.ts [hunk 26]
```diff
@@ -0,0 +1,33 @@
...
+  const contractAddresses = contractList.split(',');
+
+  const verifiers: DestinationContractHandler[] = verifierList
+    ? await getVerifiersFromArgs(verifierList, hre, 'Contract')
+    : await getVerifiersFromFile(hre, 'Contract');
+
+  for (const contractAddress of contractAddresses) {
+    for (const verifier of verifiers) {
+      try {
+        await verifier.acceptContract(contractAddress);
```
Comment (antomor) [inline tasks/destinationContractHandler/allowContracts.ts:23]: I know we applied the same pattern in other parts, but I'd do something like:
```
const tx = await verifier.acceptContract(contractAddress);
console.log(`Sent transaction ${tx.hash}`);
```

File: tasks/destinationContractHandler/allowContracts.ts [hunk 27]
```diff
@@ -0,0 +1,33 @@
...
+      } catch (error) {
+        console.error(
+          `Error adding contract with address ${contractAddress} to allowed contracts on Verifier at ${verifier.address}`
+        );
+        throw error;
+      }
+    }
+  }
+  console.log('Contracts allowed successfully!');
```
Comment (antomor) [inline tasks/destinationContractHandler/allowContracts.ts:32]: I know we applied the same pattern elsewhere, but I'd remove this

File: tasks/destinationContractHandler/removeContracts.ts [hunk 28]
```diff
@@ -0,0 +1,38 @@
...
+        const index = (await verifier.getAcceptedContracts()).indexOf(
+          contractAddress
+        );
+        if (index === -1) {
+          console.log(
+            `Contract with address ${contractAddress} is not accepted on Verifier at ${verifier.address}`
+          );
+          continue;
+        }
+        await verifier.removeContract(contractAddress, index);
```
Comment (antomor) [inline tasks/destinationContractHandler/removeContracts.ts:28]: I'd replace this with:
```
const tx = await verifier.removeContract(contractAddress);
console.log(`Sent transaction ${tx.hash}`);
```

File: tasks/destinationContractHandler/removeContracts.ts [hunk 29]
```diff
@@ -0,0 +1,38 @@
...
+      } catch (error) {
+        console.error(
+          `Error removing contract with address ${contractAddress} from allowed contracts on Verifier at ${verifier.address}`
+        );
+        throw error;
+      }
+    }
+  }
+  console.log('Contracts removed successfully!');
```
Comment (antomor) [inline tasks/destinationContractHandler/removeContracts.ts:37]: I'd remove this

File: tasks/utils.ts [hunk 30]
```diff
@@ -159,31 +176,51 @@ export const getVerifiersFromFile = async (hre: HardhatRuntimeEnvironment) => {
     nativeHolderRelayVerifier,
     boltzDeployVerifier,
     boltzRelayVerifier,
+    minimalBoltzDeployVerifier,
   } = await getVerifiers(hre);
 
+  if (type === 'Token') {
```
Comment (antomor) [inline tasks/utils.ts:182]: here I'd something like:
```
const verifiers: Record<Handler, T[]> = {
    'Token': [
      deployVerifier,
      relayVerifier,
      customDeployVerifier,
      customRelayVerifier,
      nativeHolderDeployVerifier,
      nativeHolderRelayVerifier,
      boltzDeployVerifier,
      boltzRelayVerifier,
    ] as T[],
    'Contract':[
      boltzDeployVerifier,
      boltzRelayVerifier,
      minimalBoltzDeployVerifier,
    ] as T[]
  }
  return verifiers[type];
```

File: tasks/utils.ts [hunk 31]
```diff
@@ -159,31 +176,51 @@ export const getVerifiersFromFile = async (hre: HardhatRuntimeEnvironment) => {
...
     verifierList
       .split(',')
-      .map((address) => getTokenHandlerFromAddress(address, hre))
-  );
+      .map((address) =>
+        type === 'Token'
+          ? getTokenHandlerFromAddress(address, hre)
+          : getDestinationContractHandlerFromAddress(address, hre)
+      )
+  ) as Promise<T[]>;
```
Comment (antomor) [inline tasks/utils.ts:226]: I'd replace it with something like:
```
const verifierGetters: Record<Handler, (address: string) => Promise<T>> = {
    'Token': (address) => getTokenHandlerFromAddress(address, hre) as Promise<T>,
    'Contract': (address) => getDestinationContractHandlerFromAddress(address, hre),
  };
  const verifierGetter = verifierGetters[type];

  return Promise.all(
    verifierList
      .split(',')
      .map(verifierGetter)
  ) as Promise<T[]>;
```

**Review by antomor — CHANGES_REQUESTED** [review]
@franciscotobar Thank you for the changes 🙏🏼 
I haven't tried to run it locally, but I've left a few comments. I'll test everything before approving

File: tasks/utils.ts (hunk 31 above)
Comment (franciscotobar) [reply tasks/utils.ts:226]: updated

File: tasks/utils.ts (hunk 30 above)
Comment (franciscotobar) [reply tasks/utils.ts:182]: updated

File: tasks/destinationContractHandler/removeContracts.ts (hunk 29 above)
Comment (franciscotobar) [reply tasks/destinationContractHandler/removeContracts.ts:37]: updated

File: tasks/destinationContractHandler/removeContracts.ts (hunk 28 above)
Comment (franciscotobar) [reply tasks/destinationContractHandler/removeContracts.ts:28]: updated

File: tasks/destinationContractHandler/allowContracts.ts (hunk 27 above)
Comment (franciscotobar) [reply tasks/destinationContractHandler/allowContracts.ts:32]: updated

File: tasks/destinationContractHandler/allowContracts.ts (hunk 26 above)
Comment (franciscotobar) [reply tasks/destinationContractHandler/allowContracts.ts:23]: updated

File: hardhat.config.ts (hunk 24 above)
Comment (franciscotobar) [reply hardhat.config.ts:216]: updated

File: hardhat.config.ts (hunk 25 above)
Comment (franciscotobar) [reply hardhat.config.ts:219]: updated

File: tasks/deploy.ts
[the remark this answers is not in the document]
Comment (antomor) [reply tasks/deploy.ts:1]: Implemented

File: contracts/verifier/MinimalBoltzRelayVerifier.sol (hunk 21 above)
[the remark this answers is not in the document]
Comment (antomor) [reply contracts/verifier/MinimalBoltzRelayVerifier.sol]: I've removed that code check and the related test

**Review by antomor — APPROVED** [review]

**Review by raullaprida — CHANGES_REQUESTED** [review]
I think the destinationContractHandler comment is important so I left some parts from the review out to focus on this

File: contracts/DestinationContractHandler.sol [hunk 32]
```diff
@@ -0,0 +1,53 @@
...
+
+abstract contract DestinationContractHandler is Ownable {
+    mapping(address => bool) public contracts;
+    address[] public acceptedContracts;
+
+    function acceptContract(address destinationContract) external onlyOwner {
+        require(
+            destinationContract != address(0),
+            "Contract cannot be zero address"
+        );
```
[the remark this answers is not in the document]
Comment (franciscotobar) [reply contracts/DestinationContractHandler.sol:15]: The only one that can add a new address is the owner of the contract. We are assuming that prior to adding the new contract, he knows what he is adding.  But we can also include the validation.

File: contracts/DestinationContractHandler.sol [hunk 33]
```diff
@@ -0,0 +1,53 @@
...
+        require(
+            !contracts[destinationContract],
+            "Contract is already accepted"
+        );
+        contracts[destinationContract] = true;
+        acceptedContracts.push(destinationContract);
```
[the remark this answers is not in the document]
Comment (franciscotobar) [reply contracts/DestinationContractHandler.sol:25]: updated

File: contracts/DestinationContractHandler.sol (hunk 32 above)
[the remark this answers is not in the document]
Comment (raullaprida) [reply contracts/DestinationContractHandler.sol:19]: Having the validation makes it more robust, but if you have a validation on the API code that ultimately calls this function then it would be equivalent and adding the on-chain validation a bit redundant. If the owner doesn't use the API and calls the contract directly, then you can leave a comment in the contract about the expected prerequisites of the destinationContract address.

[the remark this answers is not in the document]
Comment (franciscotobar) [reply contracts/DestinationContractHandler.sol:20]: A validation was included to make it more robust.

File: contracts/verifier/BoltzDeployVerifier.sol [hunk 34]
```diff
@@ -0,0 +1,121 @@
...
+
+/**
+ * A Verifier to be used on deploys.
+ */
+contract BoltzDeployVerifier is
+    IDeployVerifier,
+    TokenHandler,
+    DestinationContractHandler
+{
+    address private _factory;
```
[the remark this answers is not in the document]
Comment (franciscotobar) [reply contracts/verifier/BoltzDeployVerifier.sol]: updated

File: contracts/verifier/MinimalBoltzDeployVerifier.sol [hunk 35]
```diff
@@ -0,0 +1,105 @@
...
+import "../interfaces/BoltzVerifier.sol";
+
+/**
+ * A Verifier to be used on deploys.
+ */
+contract MinimalBoltzDeployVerifier is
+    IDeployVerifier,
+    DestinationContractHandler
+{
+    address private _factory;
```
[the remark this answers is not in the document]
Comment (franciscotobar) [reply contracts/verifier/MinimalBoltzDeployVerifier.sol]: updated

File: contracts/verifier/MinimalBoltzRelayVerifier.sol [hunk 36]
```diff
@@ -0,0 +1,44 @@
...
+/* solhint-disable no-inline-assembly */
+/* solhint-disable avoid-low-level-calls */
+
+/**
+ * A verifier for relay transactions.
+ */
+contract MinimalBoltzRelayVerifier is IRelayVerifier {
+    using SafeMath for uint256;
+
+    address private _factory;
```
[the remark this answers is not in the document]
Comment (franciscotobar) [reply contracts/verifier/MinimalBoltzRelayVerifier.sol]: updated

**Review by antomor — APPROVED** [review]
LGTM
